### DISCUSSION TITLE
Add constraint on parameters search

### DISCUSSION THREAD

**Discussion by qaomia** [optuna/optuna#3813] [Q&A] [answered]
Hello,

I'm trying to use optuna to optimize weights of a formula, let's say _score = **x**a + **y**b + **z**c_ with _a, b, c_ some constants, and **x, y, z** my weights.

All weights should be floats between 0 and 1, but I would like also to add the constraint **x + y + z = 1**. How could I do that?
I've been trying to follow [this kind of example](https://github.com/optuna/optuna-examples/blob/main/multi_objective/botorch_simple.py), but the constraints seem to be less or equal to 0, whereas I need exactly equal to 0.

So far, I have tried to define my parameters as follow:

```python
import optuna

def objective(trial):
    y = trial.suggest_float("y", 0, 1)
    z = trial.suggest_float("z", 0, 1)
    # define x as x + y + z = 1
    x = 1 - (y + z)

    print(x, y, z)

    # define constraint as x >= 0
    c0 = -x
    
    trial.set_user_attr("constraint", [c0])
    
    f1 = 5*x + 20*y + 3*z
    return f1

def constraints(trial):
    return trial.user_attrs["constraint"]


sampler = optuna.integration.BoTorchSampler(
        constraints_func=constraints,
        n_startup_trials=10,
    )
study = optuna.create_study(
    directions=["minimize"],
    sampler=sampler,
)
study.optimize(objective, n_trials=100)

print("Best trial:", study.best_trial)
print("Best params:", study.best_params)
```

As a result, x + y + z = 1 but x can be negative (constraint violated). Did I understand incorrectly the usage of constraints with BoTorch?

Thank you!

**Comment by nzw0301** [comment, marked as answer]
The possible approach is to set the constraint as `trial.set_user_attr("constraint", [x, -x])` since constraint value = `0` is considered feasible. However, the y and z are float values, so I think it is almost impossible to achieve that x is exactly 0 due to numerical precision. I think to solve such a hard constraint problem, Optuna's constraint optimisation is not suitable yet.

**Reply by nzw0301** [reply]
For example, https://docs.scipy.org/doc/scipy/tutorial/optimize.html#constrained-minimization-of-multivariate-scalar-functions-minimize looks more suitable package I suppose.

**Reply by gimmefriedchicken** [reply]
wouldn't this definition of constraints lead to x converging towards 0 whereas it could take any value in the range [0, 1]?

**Reply by nzw0301** [reply]
I think the solution of parameters never converges such a solution since this is a single objective optimisation; optuna tries to find single optima, not any params in specific ranges; and the constraint is not satisfied as I said in practice in this setting.

**Comment by twv20** [comment]
Another approach is to use variables to describe the limits of other variables. For example, 

```python
def objective(trial):
        y = trial.suggest_float("y", 0, 1)
        z = trial.suggest_float("z", 0, 1-y)
        # define x as x + y + z = 1
        x = 1 - (y + z)
        f1 = 5*x + 20*y + 3*z
        return f1
```

**Reply by oll-korrekt** [reply]
Warning: this will have the potentially undesirable effect of modifying the joint distribution of the points (x, y, z). In particular, z will be sampled with higher probability nearer to 0.

The only true way to solve this with a hard constraint using Optuna would be to transform to (x, y, 1 - x - y) to (x', y') into the right triangle (0,0)-(1,0)-(0,1), then reflect the right triangle to a square using `abs` so you can sample both x' and y' from 0 and 1. Your trial should suggest x' and y', and your objective function would have to undo these two transformations to use the original x, y and z.

**Reply by rodrigoscoelho** [reply]
What do you think about the following FAQ about this?
https://optuna.readthedocs.io/en/stable/faq.html#how-do-i-suggest-variables-which-represent-the-proportion-that-is-are-in-accordance-with-dirichlet-distribution
